src/Compiler/Checking/PostInferenceChecks.fs

```diff
@@ -302,11 +302,16 @@ let BindVal cenv env (v: Val) =
     //printfn "binding %s..." v.DisplayName
     let alreadyDone = cenv.boundVals.ContainsKey v.Stamp
     cenv.boundVals[v.Stamp] <- 1
+    
+    let notTopLevel () =
+        let parentHasSignature = v.DeclaringEntity.SigRange.FileName.EndsWith(".fsi")
+        not v.IsCompiledAsTopLevel || (parentHasSignature && v.SigRange.FileName.EndsWith(".fs"))
+    
     if not env.external &&
        not alreadyDone &&
        cenv.reportErrors && 
        not v.HasBeenReferenced && 
-       not v.IsCompiledAsTopLevel && 
+       notTopLevel () && 
```

Comment:
Maybe it's better to somehow reuse `FSharpSymbolUse.IsPrivateToFile` logic? It seems it can be extracted into some common code.

Reply:
It is now slightly better I guess. I'm still all ears to hear how we could improve it if possible.